internal/builder/dynamic_mappings.go

```diff
@@ -0,0 +1,256 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package builder
+
+import (
+	_ "embed"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+
+	"github.com/elastic/elastic-package/internal/formatter"
+	"github.com/elastic/elastic-package/internal/logger"
+	"github.com/elastic/elastic-package/internal/packages"
+	"github.com/elastic/elastic-package/internal/packages/buildmanifest"
+)
+
+//go:embed _static/ecs_mappings.json
+var staticEcsMappings string
+
+type ecsTemplates struct {
+	Mappings struct {
+		Properties       map[string]interface{}   `yaml:"properties"`
+		DynamicTemplates []map[string]interface{} `yaml:"dynamic_templates"`
+	} `yaml:"mappings"`
+}
+
+func addDynamicMappings(packageRoot, destinationDir string) error {
+	bm, ok, err := buildmanifest.ReadBuildManifest(packageRoot)
+	if err != nil {
+		return errors.Wrap(err, "can't read build manifest")
+	}
+	if !ok {
+		logger.Debugf("Build manifest hasn't been defined for the package")
+		return nil
+	}
+	if !bm.ImportCommonDynamicMappings() {
+		logger.Debugf("Package doesn't have to import common dynamic mappings")
+		return nil
+	}
+
+	dataStreamManifests, err := filepath.Glob(filepath.Join(destinationDir, "data_stream", "*", packages.DataStreamManifestFile))
+	if err != nil {
+		return err
+	}
+
+	for _, datastream := range dataStreamManifests {
+		contents, err := addDynamicMappingElements(datastream)
+		if err != nil {
+			return err
+		}
+		err = os.WriteFile(datastream, contents, 0664)
+		if err != nil {
+			return err
+		}
+	}
+
+	packageManifest := filepath.Join(destinationDir, packages.PackageManifestFile)
+
+	m, err := packages.ReadPackageManifest(packageManifest)
+	if err != nil {
+		return err
+	}
+	if m.Type == "input" {
+		contents, err := addDynamicMappingElements(packageManifest)
+		if err != nil {
+			return err
+		}
+		os.WriteFile(packageManifest, contents, 0664)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func addDynamicMappingElements(path string) ([]byte, error) {
+	var ecsMappings ecsTemplates
+	err := yaml.Unmarshal([]byte(staticEcsMappings), &ecsMappings)
+	if err != nil {
+		return nil, err
+	}
+
+	contents, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var doc yaml.Node
+	err = yaml.Unmarshal(contents, &doc)
+	if err != nil {
+		return nil, err
+	}
+
+	err = addEcsMappings(&doc, ecsMappings)
+	if err != nil {
+		return nil, err
+	}
+
+	err = addEcsMappingsListMeta(&doc, ecsMappings)
+	if err != nil {
+		return nil, err
+	}
+
+	contents, err = formatResult(&doc)
+	if err != nil {
+		logger.Errorf("Error formatting %s", err)
+		return nil, err
+	}
+
+	logger.Debugf("New Contents manifest:\n%s", contents)
```

Comment:
Nit. This may be too verbose, this content can be also seen by looking at the built package, right?

Reply:
mmm this was just intended for debugging purposes. Just a leftover, I'll remove it.